### ISSUE TITLE
Check for ME presence in coreboot images with ifdtool if flashmap does not indicate it

### ISSUE THREAD

**Issue by miczyg1** [Dasharo/dasharo-issues#461]
### The problem you're addressing (if any)

The utility assumes the gap from offset 0 to offset of the start of BIOS flashmap region is just closed-source without knowing what is inside. 

### Describe the solution you'd like

We could implement a usage of ifdtool to try extracting some info about flash descriptor and ME from the image.

### Where is the value to a user, and who might that user be?

1. Better comparison between vendor and open binaries.
2. More precise parsing of coreboot images.

### Describe alternatives you've considered

None

### Additional context

https://github.com/Dasharo/dasharo-issues/issues/43#issuecomment-1575335086
https://github.com/Dasharo/dasharo-issues/issues/43#issuecomment-1588776562

**Comment by miczyg1** [comment]
> [@miczyg1](https://github.com/miczyg1) Are there cases where fmap does not show ME when it exists? I ran the tool for MSI and Novacustom binaries and they both showed the ME regions, e.g:
> 
> ```
> | FMAP region | Offset | Size | Category |
> | ----------- | ------ | ---- | -------- |
> | SI_ME | 0x1000 | 0x3ff000 | closed-source |
> ```

Yes, there can be a case where the FMAP doesn't show the ME nor IFD regions. Our firmware always have these region because we explicitly define them in FMAP layouts. By default (if the board code does not provide FMAP layout file) the FMAP layout does not contain anything else than coreboot: https://github.com/coreboot/coreboot/blob/main/util/cbfstool/default-x86.fmd

**Comment by PLangowski** [comment]
@miczyg1 I see. Where can I find some binaries without those regions in the FMAP? Also, am I right to assume that those regions are all closed-source and using ifdtool won't change the openness score of binaries?

**Comment by PLangowski** [comment]
I'm mostly interested in a binary that has ME, but it won't be detected by cbfstool

**Comment by miczyg1** [comment]
> > Hmm that's not possible. coreboot build system/cbfstool creates the regions automatically when ME and IFD is provided
> 
> Then what is the goal of this issue? I thought we wanted to use `ifdtool` in case those regions exist, but aren't shown in the flash map

```
The problem you're addressing (if any)
The utility assumes the gap from offset 0 to offset of the start of BIOS flashmap region is just closed-source without knowing what is inside.

Describe the solution you'd like
We could implement a usage of ifdtool to try extracting some info about flash descriptor and ME from the image.
```

That is precisely the goal - not to assume that everything is closed source. You are mistaking ifdtool purpose with flashmap purpose. Let me clear that up a bit. This is a fragment of the output of ifdtool consuming a Dasharo image compatible with ODROID H4:

```
ifdtool -p adl -d build/coreboot.rom 
File build/coreboot.rom is 16777216 bytes
PCH Revision: 500 series Tiger Point/ 600 series Alder Point

...

Found Region Section
FLREG0:    0x00000000
  Flash Region 0 (Flash Descriptor): 00000000 - 00000fff 
FLREG1:    0x0fff0600
  Flash Region 1 (BIOS): 00600000 - 00ffffff 
FLREG2:    0x04130001
  Flash Region 2 (Intel ME): 00001000 - 00413fff 
FLREG3:    0x00007fff
  Flash Region 3 (GbE): 07fff000 - 00000fff (unused)
FLREG4:    0x00007fff
  Flash Region 4 (Platform Data): 07fff000 - 00000fff (unused)
FLREG5:    0x00007fff
  Flash Region 5 (Device Exp1): 07fff000 - 00000fff (unused)
FLREG6:    0x00007fff
  Flash Region 6 (Secondary BIOS): 07fff000 - 00000fff (unused)
FLREG8:    0x00007fff
  Flash Region 8 (EC): 07fff000 - 00000fff (unused)
FLREG9:    0x05ff0414
  Flash Region 9 (Device Exp2): 00414000 - 005fffff 
FLREG10:    0x00007fff
  Flash Region 10 (IE): 07fff000 - 00000fff (unused)
FLREG11:    0x00007fff
  Flash Region 11 (10GbE_0): 07fff000 - 00000fff (unused)
FLREG12:    0x00007fff
  Flash Region 12 (10GbE_1): 07fff000 - 00000fff (unused)
FLREG15:    0x00007fff
  Flash Region 15 (PTT): 07fff000 - 00000fff (unused)

...
```

What you need to do is to parse the `Found Region Section` fragment and extract all regions (except BIOS) that are not marked as `unused` and classify them into one of the 4 categories:

- open-source
- closed-source
- data
- empty space

Some of these regions will always contain data or code or empty space. You have the address ranges which you have to use to calculate the size of each region. Flashmap wouldn't necessarily tell you the information about these regions.

Now what the classification should look like:

- open-source - none of the regions should have any open-source code, except BIOS, but it is handled in the utility already), not sure what to do with Secondary BIOS right now, becasue I haven't seen any image with that region enabled, but it is likely that it will contain BIOS so the utility should treat it as another BIOS image and classify it according to the existing rules
- closed-source - ME, IE, PPT (unsure about this one, but let's assume closed source),  10GbE_0/10GbE_1 (Intel documentation says it is code and data, unlike in standard GbE), EC (EC region may have open code but it depends on platform and there is no easy way to determine it, on servers it may contain BMC code as well)
- data - ME, IE, GbE, Flash Descriptor, Platform Data
- empty space - ME, IE, Device Exp1/Device Exp2 (but the code should check if these regions are really empty, all 00's or all FF's)

ME/IE is in 3 categories because it may contain data, code and empty space. However, proper classification may be problematic right now, so we can simplify to be everything closed-source, for now.

**Comment by PLangowski** [comment]
@miczyg1 I see the tool marks many regions as `unused` and doesn't show the ranges. I'm not sure if this is very good. Maybe we should use `ifdtool` and ask the user to provide the microarchitecture? We could provide a table in the README, which will map some platforms to the arhitecture that needs to be passed.

**Comment by pietrushnic** [comment]
No. We had that discussion somewhere with concussion that it doesn't make sense to support it, although parser would be accepted upstream most likely, but now it doesn't work.

**Comment by PLangowski** [comment]
Then I guess we have to use either `ifdtool` or `ich_descriptors_tool`

**Comment by miczyg1** [comment]
> Will UEFITool work with Dasharo firmware?

Like @pietrushnic said, it won't work to parse anything coreboot-specific in the BIOS region. But it can be sufficiently used to parse ME firmware regions, even with coreboot images.

> Then I guess we have to use either ifdtool or ich_descriptors_tool

From those two I would choose ifdtool, becasue ich_descriptors_tool is unreliable in detecting the microarchitecture. Although we can empirically test the ich_descriptors_tool on all our Dasharo images we have and see if it is wrong somewhere. If it gives correct information about all regions for all images, then it can be used instead of ifdtool, to avoid passing extra parameter to the utility.

**Comment by PLangowski** [comment]
@miczyg1 I see many ideas here. I would like to start working on one of them. 

> The utility should just do its thing. Probably should be passed as parameter rather than asking.

Yes, that's what I meant. We can add a parameter like `--microarchitecture` to the tool and then use `ifdtool` to parse the regions. From what I understand, we should discard `unused` regions.

Is this approach acceptable? Can I start working on this?

**Comment by miczyg1** [comment]
> [@miczyg1](https://github.com/miczyg1) I see many ideas here. I would like to start working on one of them.
> 
> > The utility should just do its thing. Probably should be passed as parameter rather than asking.
> 
> Yes, that's what I meant. We can add a parameter like `--microarchitecture` to the tool and then use `ifdtool` to parse the regions. From what I understand, we should discard `unused` regions.
> 
> Is this approach acceptable? Can I start working on this?

Okay let's start with this. If we will not be pleased with results, we can always change it in the future

**Comment by PLangowski** [comment]
@miczyg1 

> empty space - ME, IE, Device Exp1/Device Exp2 (but the code should check if these regions are really empty, all 00's or all FF's)

How should we classify the regions (Device Exp) if they're not empty?

**Comment by PLangowski** [comment]
Another question - From what I understand, right now the tool classifies a lot of these regions as closed-source. Aside from parsing and classifying them using `ifdtool`, we should remove those regions from the classified region maps (`data_regions`, `code_regions`, etc.) to avoid duplication. Alternatively, we could just not parse them in the `_classify_region` function at all. Are those assumptions correct?

**Comment by PLangowski** [comment]
@miczyg1 
Just to be clear - the regions parsed with `ifdtool` will only correspond to `SI_ME` and `SI_DESC` from FMAP, so those should be deleted? Or are there any other FMAP regions that should be inspected?
